? .git
? .gitignore
? 727332-1_conditional_batching.patch
? libraries/simplepie.inc
Index: feeds.module
===================================================================
RCS file: /cvs/drupal-contrib/contributions/modules/feeds/feeds.module,v
retrieving revision 1.34
diff -u -p -r1.34 feeds.module
--- feeds.module	27 Feb 2010 02:03:34 -0000	1.34
+++ feeds.module	27 Feb 2010 02:11:30 -0000
@@ -19,6 +19,9 @@ define('FEEDS_EXPIRE_NEVER', -1);
 define('FEEDS_EXPORT_NONE', 0x0);
 // The Drupal Queue FeedsScheduler may use for scheduling importing or expiry.
 define('FEEDS_SCHEDULER_QUEUE', 'feeds_queue');
+// Status of batched operations.
+define('FEEDS_BATCH_COMPLETE', 1);
+define('FEEDS_BATCH_ACTIVE', 0);
 
 
 /**
@@ -387,7 +390,8 @@ function feeds_scheduler_work($feed_info
  */
 
 /**
- * Batch helper.
+ * Batch helper. Invokes $method on this page load. If it does not finish,
+ * starts batch processing.
  *
  * @param $title
  *   Title to show to user when executing batch.
@@ -400,14 +404,16 @@ function feeds_scheduler_work($feed_info
  *   source to be imported.
  */
 function feeds_batch_set($title, $method, $importer_id, $feed_nid = 0) {
-  $batch = array(
-    'title' => $title,
-    'operations' => array(
-      array('feeds_batch', array($method, $importer_id, $feed_nid)),
-    ),
-    'progress_message' => '',
-  );
-  batch_set($batch);
+  if (FEEDS_BATCH_COMPLETE != feeds_source($importer_id, $feed_nid)->$method()) {
+    $batch = array(
+      'title' => $title,
+      'operations' => array(
+        array('feeds_batch', array($method, $importer_id, $feed_nid)),
+      ),
+      'progress_message' => '',
+    );
+    batch_set($batch);
+  }
 }
 
 /**
Index: includes/FeedsImporter.inc
===================================================================
RCS file: /cvs/drupal-contrib/contributions/modules/feeds/includes/FeedsImporter.inc,v
retrieving revision 1.14
diff -u -p -r1.14 FeedsImporter.inc
--- includes/FeedsImporter.inc	23 Feb 2010 23:57:26 -0000	1.14
+++ includes/FeedsImporter.inc	27 Feb 2010 02:11:30 -0000
@@ -11,10 +11,6 @@ require_once(dirname(__FILE__) .'/FeedsC
 require_once(dirname(__FILE__) .'/FeedsSource.inc');
 require_once(dirname(__FILE__) .'/FeedsBatch.inc');
 
-// Status of batched operations.
-define('FEEDS_BATCH_COMPLETE', 1);
-define('FEEDS_BATCH_ACTIVE', 0);
-
 /**
  * A FeedsImporter object describes how an external source should be fetched,
  * parsed and processed. Feeds can manage an arbitrary amount of importers.
